File: compiler-core/src/main/java/com/siimkinks/sqlitemagic/structure/DatabaseStructure.java
package com.siimkinks.sqlitemagic.structure;

import com.siimkinks.sqlitemagic.Environment;
import com.siimkinks.sqlitemagic.element.ColumnElement;
import com.siimkinks.sqlitemagic.element.IndexElement;
import com.siimkinks.sqlitemagic.element.TableElement;
import com.siimkinks.sqlitemagic.processing.GenClassesManagerStep;
import com.siimkinks.sqlitemagic.util.TopsortTables;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

@Data
@ToString
@NoArgsConstructor
@AllArgsConstructor(access = AccessLevel.PRIVATE)
public final class DatabaseStructure implements Serializable {
  LinkedHashMap<String, TableStructure> tables = new LinkedHashMap<>();
  LinkedHashMap<String, IndexStructure> indices = new LinkedHashMap<>();

  public static DatabaseStructure create(Environment environment, GenClassesManagerStep managerStep) {
    final List<TableElement> allTableElements = environment.getAllTableElements();
    final LinkedHashMap<String, TableStructure> tables = new LinkedHashMap<>(allTableElements.size());
    for (TableElement tableElement : TopsortTables.sort(environment, allTableElements)) {
      final List<ColumnElement> allColumns = tableElement.getAllColumns();
      final ArrayList<ColumnStructure> columns = new ArrayList<>(allColumns.size());
      for (ColumnElement columnElement : allColumns) {
        columns.add(ColumnStructure.create(columnElement));
      }
      tables.put(tableElement.getTableName(), TableStructure.create(tableElement, columns));
    }

    final List<IndexElement> allIndexElements = managerStep.getAllIndexElements();
    final LinkedHashMap<String, IndexStructure> indices = new LinkedHashMap<>(allIndexElements.size());
    for (IndexElement indexElement : allIndexElements) {
      indices.put(indexElement.getIndexName(), IndexStructure.create(indexElement));
    }

    return new DatabaseStructure(tables, indices);
  }

  public DatabaseStructure plus(DatabaseStructure other) {
    final LinkedHashMap<String, TableStructure> tables = new LinkedHashMap<>(getTables());
    tables.putAll(other.getTables());

    final LinkedHashMap<String, IndexStructure> indices = new LinkedHashMap<>(getIndices());
    indices.putAll(other.getIndices());

    return new DatabaseStructure(tables, indices);
  }
}

File: compiler-ksp/src/main/kotlin/com/siimkinks/sqlitemagic/manager/DatabaseStructure.kt
package com.siimkinks.sqlitemagic.manager

import com.siimkinks.sqlitemagic.index.IndexElement
import com.siimkinks.sqlitemagic.model.ColumnElement
import com.siimkinks.sqlitemagic.model.TableElement
import com.siimkinks.sqlitemagic.model.schemaSql
import com.siimkinks.sqlitemagic.schema.SqliteSchema.MAIN
import com.siimkinks.sqlitemagic.schema.SqliteSchema.TEMPORARY
import com.siimkinks.sqlitemagic.view.ViewElement
import kotlinx.serialization.Serializable

@Serializable
data class ColumnStructure(
  val id: Boolean = false,
  val autoIncrement: Boolean = false,
  val name: String = "",
  val onDeleteCascade: Boolean = false,
  val sqlType: String = "",
  val schema: String = ""
) {
  companion object {
    fun from(column: ColumnElement) = with(column) {
      ColumnStructure(
        id = isId,
        autoIncrement = id?.isAutoIncrement == true,
        name = columnName,
        onDeleteCascade = relationship?.onDeleteCascade == true,
        sqlType = sqlStorageType.affinity.name,
        schema = schemaSql()
      )
    }
  }
}

@Serializable
data class TableStructure(
  val name: String = "",
  val schema: String = "",
  val columns: List<ColumnStructure> = emptyList()
) {
  companion object {
    fun from(table: TableElement) = with(table) {
      val columns = allColumns.map(transform = ColumnStructure::from)
      TableStructure(
        name = tableName,
        schema = schemaSql(
          columnSchemas = columns.map(ColumnStructure::schema)
        ),
        columns = columns
      )
    }
  }
}

@Serializable
data class IndexStructure(
  val name: String = "",
  val indexSql: String = "",
  val forTable: String = ""
) {
  companion object {
    fun from(index: IndexElement) = IndexStructure(
      name = index.name,
      indexSql = index.createSql(),
      forTable = index.tableName
    )
  }
}

@Serializable
data class ViewStructure(
  val name: String = "",
  val moduleName: String? = null
)

@Serializable
data class DatabaseStructure(
  val tables: Map<String, TableStructure> = emptyMap(),
  val indices: Map<String, IndexStructure> = emptyMap(),
  val temporaryTables: Map<String, TableStructure> = emptyMap(),
  val temporaryIndices: Map<String, IndexStructure> = emptyMap(),
  val views: Map<String, ViewStructure> = emptyMap(),
  val temporaryViews: Map<String, ViewStructure> = emptyMap()
) {
  companion object {
    internal fun from(
      orderedTables: CreationOrderedTables,
      indexes: Iterable<IndexElement> = emptyList(),
      views: Iterable<ViewElement> = emptyList()
    ) = with(orderedTables) {
      val orderedIndexes = sortedIndexes(indexes)
      val orderedViews = views.sortedBy(ViewElement::declarationOrder)
      DatabaseStructure(
        tables = persistent.associateByTo(
          destination = linkedMapOf(),
          keySelector = TableElement::tableName,
          valueTransform = TableStructure::from
        ),
        indices = orderedIndexes
          .filter { it.schema == MAIN }
          .associateByTo(
            destination = linkedMapOf(),
            keySelector = IndexElement::name,
            valueTransform = IndexStructure::from
          ),
        temporaryTables = temporary.associateByTo(
          destination = linkedMapOf(),
          keySelector = TableElement::tableName,
          valueTransform = TableStructure::from
        ),
        temporaryIndices = orderedIndexes
          .filter { it.schema == TEMPORARY }
          .associateByTo(
            destination = linkedMapOf(),
            keySelector = IndexElement::name,
            valueTransform = IndexStructure::from
          ),
        views = orderedViews
          .filter { it.schema == MAIN }
          .associate { view ->
            view.viewName to ViewStructure(
              name = view.viewName,
              moduleName = view.moduleName
            )
          },
        temporaryViews = orderedViews
          .filter { it.schema == TEMPORARY }
          .associate { view ->
            view.viewName to ViewStructure(
              name = view.viewName,
              moduleName = view.moduleName
            )
          }
      )
    }
  }

  fun persistentOnly() = DatabaseStructure(
    tables = tables,
    indices = indices,
    views = views
  )

  operator fun plus(other: DatabaseStructure) = DatabaseStructure(
    tables = LinkedHashMap(tables).apply { putAll(other.tables) },
    indices = LinkedHashMap(indices).apply { putAll(other.indices) },
    temporaryTables = LinkedHashMap(temporaryTables).apply { putAll(other.temporaryTables) },
    temporaryIndices = LinkedHashMap(temporaryIndices).apply { putAll(other.temporaryIndices) },
    views = LinkedHashMap(views).apply { putAll(other.views) },
    temporaryViews = LinkedHashMap(temporaryViews).apply { putAll(other.temporaryViews) }
  )
}
